payjoin-io/src/ohttp_keys.rs

```diff
@@ -0,0 +1,89 @@
+use ohttp::KeyConfig;
+use tokio::task::spawn_blocking;
+use url::Url;
+
+use crate::error::{Error, InternalError};
+
+/// Returns the ohttp keys from a payjoin server.
+///
+/// `proxy_endpoint` - The proxy the user want to use when
+/// requesting the ohttp keys from a payjoin server.
+/// `pj_endpoint` - The payjoin server endpoint to fetch the ohttp
+/// keys from.
+pub async fn fetch(proxy_endpoint: Url, pj_endpoint: &Url) -> Result<KeyConfig, Error> {
+    let ohttp_keys_url = pj_endpoint.join("/ohttp-keys")?;
+    let res = spawn_blocking(move || {
+        let proxy = match ProxyServer::new(&proxy_endpoint) {
+            Ok(p) => p,
+            Err(e) => return Err(e),
+        };
+        proxy.get(ohttp_keys_url.as_str()).call().map_err(Error::from)
+    })
+    .await?
+    .map_err(Error::from)?;
+    if res.status() != 200 {
+        return Err(Error(InternalError::Ureq(ureq::Error::Status(res.status(), res))));
+    }
+    let mut body = Vec::new();
+    let _ = res.into_reader().read_to_end(&mut body)?;
+    KeyConfig::decode(&body).map_err(Error::from)
+}
+
+// Proxy server struct.
+// This is used to make requests to the payjoin server.
+struct ProxyServer {
+    agent: ureq::Agent,
+}
+
+impl ProxyServer {
+    fn new(proxy: &Url) -> Result<Self, Error> {
+        let proxy = ureq::Proxy::new(Self::normalize_proxy_url(proxy)?)?;
+        let agent = ureq::AgentBuilder::new().proxy(proxy).build();
+        Ok(Self { agent })
+    }
+
+    fn get(&self, url: &str) -> ureq::Request { self.agent.get(url) }
+
+    // Normalize the Url to include the port for ureq. ureq has a bug
+    // which makes Proxy::new(...) use port 8080 for all input with scheme
+    // http regardless of the port included in the Url. This prevents that.
+    // https://github.com/algesten/ureq/pull/717
+    fn normalize_proxy_url(proxy: &Url) -> Result<String, Error> {
+        let host = match proxy.host_str() {
+            Some(host) => host,
+            None => return Err(Error(InternalError::ParseUrl(url::ParseError::EmptyHost))),
+        };
+        match proxy.scheme() {
+            "http" | "https" => Ok(format!("{}:{}", host, proxy.port().unwrap_or(80))),
+            _ => Ok(proxy.as_str().to_string()),
+        }
+    }
+}
```

Comment:
Not sure we need a Struct impl here, simple private `fn` implementations that return an agent may do. We also want to support a `danger-local-https` feature right away to make this crate compatible with local integration testing as well as payjoin-cli